File: App/BNKScopeField/Model/DPUEngine.swift
import Foundation
import Observation
import BNKKit

/// The DPU service wiring, read straight off the cluster.
@Observable
@MainActor
final class DPUEngine {
    private(set) var chains: [DPU.ServiceChain] = []
    private(set) var interfaces: [DPU.ServiceInterface] = []
    private(set) var loading = false
    private(set) var failure: String?

    func load(cluster: ManagedCluster) async {
        guard let client = try? cluster.client() else { return }
        loading = true
        failure = nil
        do {
            chains = try await client.serviceChains()
            interfaces = try await client.serviceInterfaces()
        } catch {
            chains = []
            interfaces = []
            failure = TelemetryEngine.brief(error)
        }
        loading = false
    }

    /// Chains sit on one node each, and the pair of nodes carry the same wiring —
    /// so grouping by node is how you see whether they actually match.
    var chainsByNode: [(node: String, chains: [DPU.ServiceChain])] {
        Dictionary(grouping: chains) { $0.spec.node ?? "unassigned" }
            .map { (node: $0.key, chains: $0.value) }
            .sorted { $0.node < $1.node }
    }

    /// Interfaces by kind, in the order traffic meets them: the wire first, then
    /// the host PFs, then the service ends inside.
    var interfacesByType: [(type: String, interfaces: [DPU.ServiceInterface])] {
        let order = ["physical": 0, "pf": 1, "service": 2]
        return Dictionary(grouping: interfaces) { $0.spec.interfaceType ?? "other" }
            .map { (type: $0.key, interfaces: $0.value.sorted { $0.interfaceName < $1.interfaceName }) }
            .sorted { (order[$0.type] ?? 9, $0.type) < (order[$1.type] ?? 9, $1.type) }
    }

    var readyChains: Int { chains.filter(\.isReady).count }
    var readyInterfaces: Int { interfaces.filter(\.isReady).count }
}

File: dart/bnk_engines/lib/src/dpu_engine.dart
import 'package:bnk_kit/bnk_kit.dart';

import 'brief.dart';
import 'cluster_store.dart';
import 'observable.dart';

/// The DPU service wiring, read straight off the cluster.
class DpuEngine extends Observable {
  List<ServiceChain> _chains = const [];
  List<ServiceInterface> _interfaces = const [];
  bool _loading = false;
  String? _failure;

  List<ServiceChain> get chains => _chains;
  List<ServiceInterface> get interfaces => _interfaces;
  bool get loading => _loading;
  String? get failure => _failure;

  Future<void> load(ManagedCluster cluster) async {
    final client = cluster.clientOrNull;
    if (client == null) return;
    _loading = true;
    _failure = null;
    notify();
    try {
      _chains = await client.serviceChains();
      _interfaces = await client.serviceInterfaces();
    } catch (e) {
      _chains = const [];
      _interfaces = const [];
      _failure = brief(e);
    }
    _loading = false;
    notify();
  }

  /// Chains sit on one node each, and the pair of nodes carry the same
  /// wiring, so grouping by node is how you see whether they actually match.
  List<({String node, List<ServiceChain> chains})> get chainsByNode {
    final groups = <String, List<ServiceChain>>{};
    for (final c in _chains) {
      groups.putIfAbsent(c.spec.node ?? 'unassigned', () => []).add(c);
    }
    final keys = groups.keys.toList()..sort();
    return [for (final k in keys) (node: k, chains: groups[k]!)];
  }

  /// Interfaces by kind, in the order traffic meets them: the wire first,
  /// then the host PFs, then the service ends inside.
  List<({String type, List<ServiceInterface> interfaces})> get interfacesByType {
    const order = {'physical': 0, 'pf': 1, 'service': 2};
    final groups = <String, List<ServiceInterface>>{};
    for (final i in _interfaces) {
      groups.putIfAbsent(i.spec.interfaceType ?? 'other', () => []).add(i);
    }
    final keys = groups.keys.toList()
      ..sort((a, b) {
        final byOrder = (order[a] ?? 9).compareTo(order[b] ?? 9);
        return byOrder != 0 ? byOrder : a.compareTo(b);
      });
    return [
      for (final k in keys)
        (type: k, interfaces: groups[k]!..sort((a, b) => a.interfaceName.compareTo(b.interfaceName)))
    ];
  }

  int get readyChains => _chains.where((c) => c.isReady).length;
  int get readyInterfaces => _interfaces.where((i) => i.isReady).length;
}
